app/services/chf/create_dzi_service.rb

```diff
@@ -0,0 +1,263 @@
+
+          http.request request do |response|
+            open local_original_file_path, 'wb' do |io|
+              response.read_body do |chunk|
+                io.write chunk
+              end
+            end
+          end
+        end
+      end
+      Rails.logger.debug("#{self.class.name}: fetch_from_fedora: #{fedora_fetch_benchmark}")
+      unless response.code == "200"
+        raise StandardError.new("Could not fetch file from fedora: #{uri}")
+      end
+    end
+
+    def create_dzi!
+      dzi_benchmark = Benchmark.measure do
+        # http://libvips.blogspot.com/2013/03/making-deepzoom-zoomify-and-google-maps.html
+        args = [vips_command, "dzsave", local_original_file_path, local_dzi_base_path, "--suffix", ".jpg[Q=#{jpeg_quality}]"]
+
+        if suppress_vips_stderr
+          args << { :err => "/dev/null" }
+        end
+
+        system(*args) or raise StandardError.new("vips dzsave failed")
+      end
+      Rails.logger.debug("#{self.class.name}: create_dzi: #{dzi_benchmark}")
+    end
+
+    # uploading to s3 is actually the slowest part if done serially.
+    # We apply a healthy dose of concurrency.
+    def upload_to_s3!
+      s_time = Time.now
+
+      # All the jpgs, which are in a _files/ dir, and subdirs of that.
+      futures = []
+      dir_path = Pathname.new(local_dzi_dir_path)
+      path_prefix_re = /\A#{Regexp.quote(WORKING_DIR.end_with?('/') ? WORKING_DIR : WORKING_DIR + '/')}/
+
+      Dir.glob("#{dir_path}/**/*.jpg").each do |full_path|
+        futures << Concurrent::Future.execute(executor: self.class.thread_pool_executor) do
+          s3_bucket.
+            object(full_path.sub(path_prefix_re, '')).
+            upload_file(full_path, acl: acl, cache_control: cache_control)
+        end
+      end
+
+      # wait on em all
+      futures.collect(&:value)
+
+      # any errors? Raise one of em.
+      if rejected = futures.find(&:rejected?)
+        raise rejected.reason
+      end
+
+      # upload .dzi AFTER all the tiles, so it's not there until they are
+      s3_bucket.
+        object(dzi_file_name).
+        upload_file(local_dzi_file_path, acl: acl, cache_control: cache_control)
+
+
+      Rails.logger.debug("#{self.class.name}: upload_to_s3: #{Time.now - s_time}")
+    end
+
+    def self.thread_pool_executor
+      @thread_pool ||= Concurrent::ThreadPoolExecutor.new(
+          min_threads:     UPLOAD_THREADS,
+          max_threads:     UPLOAD_THREADS,
+          auto_terminate:  true,
+          idletime:        60, # 1 minute. shouldn't matter, we have min and max the same
+          max_queue:       0, # unlimited
+          fallback_policy: :abort # shouldn't matter -- 0 max queue
+      )
+    end
+    self.thread_pool_executor # init now
+
+    # include the checksum so it's self-cache-busting if file at this URL
+    # changes, say, due to versioning. HOWEVER, this does make indexing
+    # somewhat slower. TODO optimize somehow. Less fetches to fedora and/or get
```

Comment:
removed. It may not be as optimized as it could be, but ti's just fighting with ActiveFedora and I don't even know.